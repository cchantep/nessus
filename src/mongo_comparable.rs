// MongoComparable provides evidence that a value of type A can be compared with a value of type B
// in MongoDB queries. This trait is used at the struct level to indicate that specific fields
// can be compared with specific types of values in MongoDB operations.
pub trait MongoComparable<A, B> {}
