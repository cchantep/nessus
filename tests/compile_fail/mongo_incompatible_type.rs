// This test verifies that MongoComparable fails with incompatible types

use tnuctipun::{FieldWitnesses, MongoComparable, HasField, FieldName};
use tnuctipun::mongo_comparable::MongoComparable as MongoComparableTrait;
use tnuctipun::HasField;
use serde::{Serialize, Deserialize};
#[derive(Debug, Clone, Serialize, Deserialize, FieldWitnesses, MongoComparable)]
struct Product {
    pub Name: String,
    pub Price: f64,
}

// Custom type that doesn't implement Into<mongodb::bson::Bson>
struct CustomType;

fn main() {
    // This should fail because CustomType doesn't implement Into<mongodb::bson::Bson>
    fn assert_implements_mongo_comparable<T, A, B>()
    where
        T: MongoComparableTrait<A, B>
    {}
    
    // Attempt to compare String field with CustomType
    assert_implements_mongo_comparable::<Product, 
        <Product as tnuctipun::field_witnesses::HasField<product_fields::Name>>::Value, 
        CustomType>();
}
