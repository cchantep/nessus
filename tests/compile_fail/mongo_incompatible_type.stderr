error[E0277]: the trait bound `Product: MongoComparable<std::string::String, CustomType>` is not satisfied
  --> tests/compile_fail/mongo_incompatible_type.rs:24:42
   |
24 |     assert_implements_mongo_comparable::<Product,
   |                                          ^^^^^^^ the trait `MongoComparable<std::string::String, CustomType>` is not implemented for `Product`
   |
   = help: the following other types implement trait `MongoComparable<A, B>`:
             `Product` implements `MongoComparable<f64, f32>`
             `Product` implements `MongoComparable<f64, f64>`
             `Product` implements `MongoComparable<f64, i16>`
             `Product` implements `MongoComparable<f64, i32>`
             `Product` implements `MongoComparable<f64, i64>`
             `Product` implements `MongoComparable<std::string::String, std::string::String>`
note: required by a bound in `assert_implements_mongo_comparable`
  --> tests/compile_fail/mongo_incompatible_type.rs:20:12
   |
18 |     fn assert_implements_mongo_comparable<T, A, B>()
   |        ---------------------------------- required by a bound in this function
19 |     where
20 |         T: MongoComparableTrait<A, B>
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_implements_mongo_comparable`
